Fix lp:1661488, Allow to enable userstat=1 for existing connections, keep concurrent connections tracking after FLUSH USER/CLIENT_STATISTICS

ihanick opened the pull request:
Bug fix for https://bugs.launchpad.net/percona-server/+bug/1661488

http://jenkins.percona.com/job/percona-server-5.5-param/1526

For already existing connections tracking required, I've added diff_connections variable for that purpose.
At the same time connections are tracked in two places (USER_STATISTICS and CLIENT_STATISTICS), thus this variable could be bigger than one for each connection.

In addition to enabling userstats in runtime, there is a FLUSH * issue. I've modified corresponding refresh code path in almost the same way as it was in original google patch.

laurynas-biveinis commented on sql/mysqld.h:
```diff
@@ -511,10 +511,13 @@ void init_global_index_stats(void);
 void init_global_client_stats(void);
 void init_global_thread_stats(void);
 void free_global_user_stats(void);
+void free_global_user_stats(void);
```
Duplicated line

laurynas-biveinis commented on sql/sql_class.h:
```diff
@@ -2347,6 +2347,7 @@ class THD :public Statement,
   ulonglong diff_commit_trans, diff_rollback_trans;
   // Number of connection errors not reflected in global_user_stats yet.
   ulonglong diff_denied_connections, diff_lost_connections;
+  ulonglong diff_connections;
```
Please add a comment

laurynas-biveinis commented on sql/sql_connect.cc:
```diff
@@ -184,14 +184,17 @@ void init_user_stats(USER_STATS *user_stats,
-
-  user_stats->user_len=               strlen(user_stats->user);
-  user_stats->priv_user_len=          strlen(user_stats->priv_user);
+  if (user)
```
Why is this check required?

laurynas-biveinis commented on sql/sql_connect.cc:
```diff
@@ -346,6 +349,68 @@ void init_global_index_stats(void)
+    return;
+  }
+
+  for (uint idx=0;idx < stats->records; idx++)
```
s/uint/ulong, space after ;

laurynas-biveinis commented on sql/sql_connect.cc:
```diff
@@ -346,6 +349,68 @@ void init_global_index_stats(void)
+  for (uint idx=0;idx < stats->records; idx++)
+  {
+    USER_STATS* user_stats=
+      (USER_STATS *) my_hash_element(stats, idx);
```
C++ style cast

laurynas-biveinis commented on sql/sql_connect.cc:
```diff
@@ -346,6 +349,68 @@ void init_global_index_stats(void)
+      (USER_STATS *) my_hash_element(stats, idx);
+    if (user_stats->concurrent_connections)
+    {
+      init_user_stats(user_stats, NULL, NULL,
```
This function has way too many zero arguments - this is not the part of this bugfix, but I have to ask you to clean it up anyway: please only keep the arguments that passed here or at the other call in increment_count_by_name, and move the rest of zero initialisation to the function body. Please also make it "static" and remove its prototype from sql/structs.h

laurynas-biveinis commented on sql/sql_connect.cc:
```diff
@@ -346,6 +349,68 @@ void init_global_index_stats(void)
+    }
+    else
+    {
+      if (insert_dynamic(&delete_items, (uchar *) &user_stats))
```
if (unlikely(insert_dynamic ..., C++ style cast

laurynas-biveinis commented on sql/sql_connect.cc:
```diff
@@ -346,6 +349,68 @@ void init_global_index_stats(void)
+{
+  DYNAMIC_ARRAY delete_items;
+
+  if (my_init_dynamic_array(&delete_items, sizeof(USER_STATS *),
```
if (unlikely(my_init ...

laurynas-biveinis commented on sql/sql_connect.cc:
```diff
@@ -346,6 +349,68 @@ void init_global_index_stats(void)
+    }
+  }
+
+  for (uint i= 0; i < delete_items.elements; ++i)
```
s/uint/ulong

laurynas-biveinis commented on sql/sql_connect.cc:
```diff
@@ -346,6 +349,68 @@ void init_global_index_stats(void)
+  for (uint i= 0; i < delete_items.elements; ++i)
+  {
+    USER_STATS *entry= *dynamic_element(&delete_items, i, USER_STATS **);
+    (void) my_hash_delete(stats, (uchar*) entry);
```
It's a bug if this my_hash_delete fails - please DBUG_ASSERT its return value; C++-style cast

laurynas-biveinis commented on sql/sql_connect.cc:
```diff
@@ -346,6 +349,68 @@ void init_global_index_stats(void)
   }
 }
 
+void reset_global_user_or_client_stats(HASH* stats)
```
static

laurynas-biveinis commented on mysql-test/include/percona_userstat_conn_handling.inc:
```diff
@@ -71,6 +71,63 @@ connection default;
+SELECT 1;
+connection default;
+
+# Verify that flush statistics keeps concurent connections counter inact
```
intact

laurynas-biveinis commented on mysql-test/include/percona_userstat_conn_handling.inc:
```diff
@@ -71,6 +71,63 @@ connection default;
 --let $assert_cond= "$query_result" = 0
 --source include/assert.inc
 
+# Make a single connection for mysqltest_1 user
```
The testcase does not actually query the value?

laurynas-biveinis commented on mysql-test/include/percona_userstat_conn_handling.inc:
```diff
@@ -71,6 +71,63 @@ connection default;
+let $count_sessions= 1;
+--source include/wait_until_count_sessions.inc
+
+# verify flush user statistics with existing connections
```
add connection default here

laurynas-biveinis replied:
This comment is not addressed yet (but connection default should be added immediately after disconnect conn1 above)

laurynas-biveinis replied:
This comment has not been addressed

laurynas-biveinis replied:
Still not addressed

laurynas-biveinis commented on mysql-test/include/percona_userstat_conn_handling.inc:
```diff
@@ -71,6 +71,63 @@ connection default;
+# Make a single connection for mysqltest_1 user
+connect (conn1,localhost,mysqltest_1,,);
+SELECT 1;
+connection default;
```
This one is redundant

ihanick commented:
http://jenkins.percona.com/job/percona-server-5.5-param/1529/
- check connection_concurrency for fault login
- check for count(1) ... where concurrency=0 instead of let $assert_cond= "$query_result" = 0
- rewrite init_user_stats, make it static

laurynas-biveinis commented on mysql-test/include/percona_userstat_conn_handling.inc:
```diff
@@ -44,33 +44,80 @@ connection default;
 --let $assert_cond= "$query_result" = 1
 --source include/assert.inc
 
-
```
Redundant change?

laurynas-biveinis replied:
This comment has not been addressed

laurynas-biveinis replied:
Still not addressed

laurynas-biveinis commented on mysql-test/include/percona_userstat_conn_handling.inc:
```diff
@@ -44,33 +44,80 @@ connection default;
 --let $assert_text= I_S.USER_STATISTICS CONCURRENT_CONNECTIONS: no connections - zero concurrency
---let $query_result= query_get_value(SELECT CONCURRENT_CONNECTIONS FROM INFORMATION_SCHEMA.USER_STATISTICS WHERE USER LIKE 'mysqltest_1', CONCURRENT_CONNECTIONS, 1)
---let $assert_cond= "$query_result" = 0
+--let $query_result= query_get_value(SELECT COUNT(1) cnt FROM INFORMATION_SCHEMA.USER_STATISTICS WHERE USER LIKE 'mysqltest_1' AND  CONCURRENT_CONNECTIONS=0, cnt, 1)
```
Why this change is required?

laurynas-biveinis replied:
OK, I see

laurynas-biveinis commented on mysql-test/include/percona_userstat_conn_handling.inc:
```diff
@@ -44,33 +44,80 @@ connection default;
+--source include/wait_until_count_sessions.inc
+
+# verify flush user statistics with existing connections
+# connection concurrency tracked only if it runs queries while userstat=1
```
No actual userstat query in this test part?

laurynas-biveinis replied:
This comment has not been addressed

laurynas-biveinis replied:
OK

laurynas-biveinis commented on mysql-test/include/percona_userstat_conn_handling.inc:
```diff
@@ -44,33 +44,80 @@ connection default;
+--source include/wait_until_count_sessions.inc
+
+
+# verify flush client statistics with existing connections
```
Likewise?

laurynas-biveinis replied:
This comment has not been addressed

laurynas-biveinis replied:
OK

laurynas-biveinis commented on mysql-test/include/percona_userstat_conn_handling.inc:
```diff
@@ -44,33 +44,80 @@ connection default;
+--let $assert_cond= "$query_result" = 1
 --source include/assert.inc
 
+# We need alive connection to check that flush will not remove concurent_connections counter
```
No actual userstat query in this test part?

laurynas-biveinis replied:
This comment has not been addressed

laurynas-biveinis commented on sql/sql_connect.cc:
```diff
@@ -158,63 +158,29 @@ void free_thread_stats(THREAD_STATS* thread_stats)
-                     ulonglong lost_connections,
-                     ulonglong access_denied_errors,
-                     ulonglong empty_queries)
+static void init_user_stats(USER_STATS *user_stats)
```
I suggested to remove only the always-zero args, not all the args. The current way is also fine, but then this whole function should be replaced by a memset.

If you choose to restore the non-zero args, then use memset, and then set only those args.

laurynas-biveinis replied:
This comment has not been addressed

laurynas-biveinis commented on sql/sql_connect.cc:
```diff
@@ -346,6 +312,60 @@ void init_global_index_stats(void)
+    if (user_stats->concurrent_connections)
+    {
+      uint saved_concurrent_connections = user_stats->concurrent_connections;
+      init_user_stats(user_stats);
```
The need for saved_concurrent_connections would disappear with init_user_stats taking a concurrent_connections arg

laurynas-biveinis replied:
See the comment at init_user_stats

laurynas-biveinis commented on sql/sql_connect.cc:
```diff
@@ -346,6 +312,60 @@ void init_global_index_stats(void)
+  for (ulong i= 0; i < delete_items.elements; ++i)
+  {
+    USER_STATS *entry= *dynamic_element(&delete_items, i, USER_STATS **);
+    DBUG_ASSERT(!my_hash_delete(stats, reinterpret_cast<uchar*>(entry)));
```
This is a bug: assert with a side effect, will get compiled away in a release build. It has to be
#ifndef DBUG_OFF
    whatervertype result=
#endif
    my_hash_delete(...);
    DBUG_ASSERT(!result);

laurynas-biveinis replied:
This comment has not been addressed

laurynas-biveinis commented:
Jenkins link is 404?

ihanick commented:
http://jenkins.percona.com/job/percona-server-5.5-param/1537/
- added DBUG_ASSERT for my_hash_delete results
- init_user_stats clears zero fields
- avoid ---let $assert_cond= "$query_result" = 0, because there is no alert for zero rows returned
- changed unknown user connection test to wrong password check, unknown users not creating any record in user stats.
- clarify flush stats test part

ihanick commented:
http://jenkins.percona.com/job/percona-server-5.5-param/1546/

- added I_S.USER_STATISTICS and I_S.CLIENT_STATISTICS related to userstats=0 <=> userstats=1 dynamic changes
- Found that it's impossible to track concurrent_connections state while userstats is disabled
- Considered to save state in the same place as in Facebook MySQL: connections are already tracked inside user resources limit
- reverted https://bugs.launchpad.net/percona-server/+bug/728082
- for each show user_statistics or client_statistics query updating number of concurrent connections  by calculating SUM(resource limit concurrent connection) for all users

Benefits:
- smaller change set (diff to original 2791a612d407e8710b690ac1d0b739005bfcec3c )
- correct tracking if user stats was enabled or disabled
- correct flush behaviour (removes all records from statistics)
- same test cases could be used
- codepath is the same for thread pool and one-thread-per-connection, I can merge both test cases to one and eliminate need in additional mysqld restart for mysql-test run

laurynas-biveinis commented on sql/sql_connect.cc:
```diff
@@ -655,6 +649,49 @@ void update_global_user_stats(THD* thd, bool create_user, time_t now)
+  {
+    USER_STATS* user_stats=
+      reinterpret_cast<USER_STATS*>(my_hash_element(stats, idx));
+    user_stats->concurrent_connections = 0;
```
No space before "="

laurynas-biveinis commented on sql/sql_connect.cc:
```diff
@@ -655,6 +649,49 @@ void update_global_user_stats(THD* thd, bool create_user, time_t now)
+
+static void inc_stats_concurrent_conn(HASH* stats, const char* user_string, int cnt) {
+  USER_STATS* user_stats;
+  if ((user_stats = (USER_STATS *) my_hash_search(stats,
```
Move assignment to var declaration (can make the pointer itself const too)
C++ style cast

laurynas-biveinis commented on sql/sql_connect.cc:
```diff
@@ -655,6 +649,49 @@ void update_global_user_stats(THD* thd, bool create_user, time_t now)
+                                                  (uchar *) user_string,
+                                                  strlen(user_string))))
+  {
+    user_stats->concurrent_connections += cnt;
```
No space before +=

laurynas-biveinis commented on sql/sql_connect.cc:
```diff
@@ -655,6 +649,49 @@ void update_global_user_stats(THD* thd, bool create_user, time_t now)
+}
+
+// Updates number of concurrent connections for user_stats and client_stats
+// based on account resoure limits
```
"resource"

laurynas-biveinis replied:
If a header comment is added (strictly speaking it has to be added for all the new functions, but the previous two are trivial enough so that's not a critical issue), then it has to be a Doxygen one

laurynas-biveinis commented on sql/sql_connect.cc:
```diff
@@ -655,6 +649,49 @@ void update_global_user_stats(THD* thd, bool create_user, time_t now)
+  clear_stats_concurrent_connections(&global_client_stats);
+  mysql_mutex_unlock(&LOCK_global_user_client_stats);
+
+  for (ulong idx=0; idx < hash_user_connections.records; idx++)
```
space after =

laurynas-biveinis commented on mysql-test/include/percona_userstat_conn_handling.inc:
```diff
@@ -1,5 +1,7 @@
 SET @userstat_old= @@userstat;
+SET @max_user_connections_old= @@max_user_connections;
 SET GLOBAL userstat=ON;
+SET GLOBAL max_user_connections = 10000;
```
Why is this required?

laurynas-biveinis commented on sql/sql_connect.cc:
```diff
@@ -655,6 +649,49 @@ void update_global_user_stats(THD* thd, bool create_user, time_t now)
+
+  for (ulong idx=0; idx < hash_user_connections.records; idx++)
+  {
+    struct user_conn * uc=
```
const

ihanick commented:
http://jenkins.percona.com/job/percona-server-5.5-param/1552/
- fixed spacing & brackets
- changed function comment to doxygen one
- const for pointers
- using per-user resource limits
- added connection default after each disconnect

laurynas-biveinis commented on mysql-test/include/percona_userstat_conn_handling.inc:
```diff
@@ -1,11 +1,16 @@
 
 # Make a single connection for mysqltest_1 user
 connect (conn1,localhost,mysqltest_1,,);
+connection conn1;
```
Redundant

laurynas-biveinis commented on mysql-test/include/percona_userstat_conn_handling.inc:
```diff
@@ -47,32 +53,106 @@ connection default;
 DROP USER mysqltest_2@localhost;
 
-SET GLOBAL userstat= @userstat_old;
+SET GLOBAL userstat= @userstat_old;
```
Missing newline

laurynas-biveinis commented on sql/sql_connect.cc:
```diff
@@ -655,6 +649,54 @@ void update_global_user_stats(THD* thd, bool create_user, time_t now)
+{
+  USER_STATS* const user_stats=
+    reinterpret_cast<USER_STATS*>(my_hash_search(stats,
+          (uchar *) user_string,
```
C++-style cast

laurynas-biveinis commented on sql/sql_connect.cc:
```diff
@@ -655,6 +649,54 @@ void update_global_user_stats(THD* thd, bool create_user, time_t now)
+
+  for (ulong idx= 0; idx < hash_user_connections.records; idx++)
+  {
+    struct user_conn* const uc=
```
const struct user_conn* const uc=

ihanick commented:
http://jenkins.percona.com/job/percona-server-5.5-param/1554/
- c++ cast (and use const uchar*)
- missing new line character at file end
- const pointer and item for userstats iteration

ihanick commented:
http://jenkins.percona.com/job/percona-server-5.5-param/1556/
There is no dependency for thread pool in current implementation,
removed thread pool test case for userstats
transferred inc file to test file as is.
